File: sys/codegen/lib/mpi_Finalize/mpi_Finalize.c
#include "mpi_Finalize.h"
#include "mpi.h"
#include "petscsys.h"
#include "petscksp.h"
#include "m2c.h"

static void b_m2c_error(int varargin_3);
static void m2c_error(int varargin_3);
static void b_m2c_error(int varargin_3)
{
  M2C_error("MPI:RuntimeError", "MPI_Finalize returned error code %d\n",
            varargin_3);
}

static void m2c_error(int varargin_3)
{
  M2C_error("MPI:RuntimeError", "MPI_Finalized returned error code %d\n",
            varargin_3);
}

void mpi_Finalize(int *info, boolean_T *toplevel)
{
  int flag;
  int b_info;
  int b_flag;
  *info = 0;
  b_info = MPI_Finalized(&flag);
  if (b_info != 0) {
    b_flag = (M2C_DEBUG);
    if (b_flag != 0) {
      m2c_error(b_info);
    }
  }

  if (!(flag != 0)) {
    *info = MPI_Finalize();
  }

  *toplevel = true;
  if (*info != 0) {
    b_m2c_error(*info);
  }
}

void mpi_Finalize_initialize(void)
{
}

void mpi_Finalize_terminate(void)
{
}

File: mpi/mpi_Finalize.m
function [info, toplevel] = mpi_Finalize
%mpi_Finalize  Terminates MPI execution environment
%
%  info =  mpi_Finalize
%
%  info(int) return code
%
%  SEE ALSO: mpi_Init, mpi_Initialized, mpi_Finalized
%
%  Note: This is a system/driver level function. A user function in general 
%  should not call this function. It is called automatically at exit
%  by MATLAB/Octave if the MMPI module was initialized.
%
% C interface:
% int MPI_Finalize(void )
% http://mpi.deino.net/mpi_functions/MPI_Finalize.html

%#codegen -args {}

info = int32(0);

if ~mpi_Finalized
    info = coder.ceval('MPI_Finalize');
end

toplevel = nargout>1;
if info && (toplevel || m2c_debug)
    m2c_error('MPI:RuntimeError', 'MPI_Finalize failed with error message %s\n', mpi_Error_string(info))
end
